### DISCUSSION TITLE
gpu-rocm check in self hosted runners GHA

### DISCUSSION THREAD

**Discussion by jimw567** [ggml-org/llama.cpp#26728] [General]
If you raised a new PR in the past couple of days, you probably have noticed a new check "gpu-rocm" in CI. If you haven't noticed it, that is good news as it means it is working as expected for your PR. 

**[CI (self-hosted) / gpu-rocm (pull_request)](https://github.com/ggml-org/llama.cpp/actions/runs/31152520487/job/92784981178?pr=26696)
CI (self-hosted) / gpu-rocm (pull_request)Successful in 10m**

The check is currently dispatched to a Strix Halo runner that I maintain. While the runner is generally working it has some issue with uploading run logs to github. I am working on resolving the issue. In the meantime, if you see any failure with gpu-rocm check and need the log, please ping me. I will help extract the log and/or debug the failure if needed.

**Comment by taronaeo** [comment]
https://github.com/ggml-org/llama.cpp/actions/runs/31421346673/job/93562566877

@jimw567 Can I have the logs for this run? Unsure why it failed.

**Reply by jimw567** [reply]
It looks like it has something to do with the newly added test test-backend-sampler.cpp from the PR below. What is the general policy on CI check failure? Would it be possible to make gpu-rocm check gating unless someone from ggml-org/amd team takes a look at it?
 
https://github.com/ggml-org/llama.cpp/pull/25532 

 n_tokens: 2
  token[0]: tok=151643, pos=0, n_seq_id=1, seq_ids=[0], logits=0
  token[1]: tok=8373 , pos=1, n_seq_id=1, seq_ids=[0], logits=1
  dist & cpu sampled id:6291, string:' solution'
  backend dist & cpu sampling test PASSED
  >>> test_backend_multi_output_limit expected error start:
  <<< test_backend_multi_output_limit expected error end.
  backend multi-output limit test PASSED
  backend multi-sequence multi-output dist test PASSED
  backend multi-output dist transaction test PASSED
  /scratch/actions-runner/_work/llama.cpp/llama.cpp/tests/test-backend-sampler.cpp:1844: GGML_ASSERT(sampled_probs != nullptr) failed
  ⚠️ warning: The current terminal doesn't support styling.  Styled output might not appear as expected.
  [New LWP 1780902]
  [New LWP 1780900]
  [New LWP 1780899]
  [Thread debugging using libthread_db enabled]
  Using host libthread_db library "/usr/lib/x86_64-linux-gnu/libthread_db.so.1".
  0x00007e0d2ccaca82 in ?? () from /usr/lib/x86_64-linux-gnu/libc.so.6
  #0  0x00007e0d2ccaca82 in ?? () from /usr/lib/x86_64-linux-gnu/libc.so.6
  #1  0x00007e0d2cca061c in ?? () from /usr/lib/x86_64-linux-gnu/libc.so.6
  #2  0x00007e0d2cd1cd4f in wait4 () from /usr/lib/x86_64-linux-gnu/libc.so.6
  #3  0x00007e0d2d50cc95 in ggml_print_backtrace () at /scratch/actions-runner/_work/llama.cpp/llama.cpp/ggml/src/ggml.c:234
  234             waitpid(child_pid, NULL, 0);
  #4  0x00007e0d2d50ce4c in ggml_abort (file=0x5de1a7d04138 "/scratch/actions-runner/_work/llama.cpp/llama.cpp/tests/test-backend-sampler.cpp",
  line=1844, fmt=0x5de1a7d04121 "GGML_ASSERT(%s) failed") at /scratch/actions-runner/_work/llama.cpp/llama.cpp/ggml/src/ggml.c:268
  268             ggml_print_backtrace();
  #5  0x00005de1a7ce66b1 in test_backend_multi_output_sampling_chain (params=...) at
  /scratch/actions-runner/_work/llama.cpp/llama.cpp/tests/test-backend-sampler.cpp:1844
  1844            GGML_ASSERT(sampled_probs != nullptr);
  #6  0x00005de1a7ce8bf1 in run_tests (tests=std::vector of length 15, capacity 16 = {...}, args=...) at
  /scratch/actions-runner/_work/llama.cpp/llama.cpp/tests/test-backend-sampler.cpp:2128
  2128                test->fn(args);
  #7  0x00005de1a7ce8eab in main (argc=1, argv=0x7ffcf688cf28) at /scratch/actions-runner/_work/llama.cpp/llama.cpp/tests/test-backend-sampler.cpp:2161
  2161            run_tests(tests, params);
  [Inferior 1 (process 1780898) detached]

      Start 44: test-state-restore-fragmented
  5/6 Test #44: test-state-restore-fragmented ....   Passed    0.55 sec
      Start 45: test-save-load-state
  6/6 Test #45: test-save-load-state .............   Passed    0.68 sec

  83% tests passed, 1 tests failed out of 6

**Reply by jimw567** [reply]
@taronaeo I just raised PR below to address this regression. Please review.

https://github.com/ggml-org/llama.cpp/pull/26878

**Reply by taronaeo** [reply]
> What is the general policy on CI check failure? Would it be possible to make gpu-rocm check gating unless someone from ggml-org/amd team takes a look at it?

I'm afraid I don't understand. Do you mean allow gpu-rocm to run only if an AMD dev approves it?

**Reply by jimw567** [reply]
Sorry about the confusion. I meant that if the gpu-rocm check fails on a PR, the PR shouldn't be merged. We can help resolve whatever issues there may be in the check. Keeping the gpu-rocm check green at all times is critical to our end users.

**Reply by taronaeo** [reply]
Oh, generally if the CI failure is related to the PR, it will not be merged until its resolved.

However if the issue appears to be consistent across many other PRs for example a HuggingFace API rate limit error, we are usually okay with merging while red.

I think the main bottleneck for us right now is the inability to read the error logs to determine if its a blocker or to safely ignore. Can we somehow expedite the firewall whitelisting on AMD's side?

**Reply by jimw567** [reply]
Totally agree that this log situation is dragging us. I followed up with our DevOps this morning. I will provide updates as I get them. In the mean time, please ping me anytime to extract logs.

**Comment by jimw567** [comment]
@taronaeo I put a temporary automation in place to upload the error log when gpu-rocm check fails. Please take a look the comment below and let me know if this helps.

https://github.com/ggml-org/llama.cpp/pull/26696#issuecomment-5255704885

**Reply by taronaeo** [reply]
That helps a lot! Thank you

**Comment by CISC** [comment]
@jimw567 Not entirely sure what's going on, but `test-recurrent-state-rollback` is failing on `gpu-rocm` currently:
https://github.com/ggml-org/llama.cpp/actions/runs/34111082341/job/101707237311#step:3:1440

**Reply by jimw567** [reply]
@CISC  Let me take a look and get back to you.
